### PR TITLE
Add support for Vue2 functional component to `vue/no-unused-properties` and `vue/no-undef-properties` rules

### LINKED ISSUE #1312: `vue/no-unused-properties` does not recognize props in functional component (Vue 2)
**Checklist**

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 7.10.0
- **eslint-plugin-vue version:** 7.0.0
- **Node version:** 12.18.4
- **Operating System:** Ubuntu 20.04

**Please show your full configuration:**

```json5
{
  "env": {
    "es6": true,
    "node": true
  },
  "parserOptions": {
    "ecmaVersion": 2017
  },
  "plugins": ["vue"],
  "extends": ["plugin:vue/recommended"],
  "rules": {
    "vue/no-unused-properties": ["error", {
      "groups": ["props", "data", "computed", "methods", "setup"]
    }]
  }
}
```

**What did you do?**	

```vue
<template functional>
  <time :datetime="props.date.toISOString()" :title="props.date.toISOString()">{{
    props.date.toISOString().replace(/T.*?$/, ``)
  }}</time>
</template>

<script>
export default {
  props: {
    date: {
      type: Date,
      required: true,
    },
  },
};
</script>
```

**What did you expect to happen?**
No error, since the `date` prop is used inside the template.

**What actually happened?**
```
error: 'date' of property found, but never used (vue/no-unused-properties) at ui/components/global/OflTime.vue:10:5:
   8 | export default {
   9 |   props: {
> 10 |     date: {
     |     ^
  11 |       type: Date,
  12 |       required: true,
  13 |     },
```

**Repository to reproduce this issue**
https://github.com/OpenLightingProject/open-fixture-library/blob/45508de7c812e9e4d9a55c9517d969c6606d53fd/ui/components/global/OflTime.vue

**Related issue:** #1669

### LINKED ISSUE #1669: `vue/no-undef-properties` reports invalid issues in functional components (Vue 2)
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 7.32.0
- **eslint-plugin-vue version:** 7.20.0
- **Node version:** 14.18.1
- **Operating System:** Ubuntu 20.04

**Please show your full configuration:**

```json5
{
  "env": {
    "es6": true,
    "node": true,
  },
  "parserOptions": {
    "ecmaVersion": 2021,
  },
  "plugins": ["vue"],
  "extends": ["eslint:recommended", "plugin:vue/recommended"],
  "rules": {
    "vue/no-undef-properties": "error"
  }
}
```

**What did you do?**	

```vue
<template functional>
  <section
    :ref="data.ref"
    :class="[data.class, data.staticClass, props.name]"
    :style="[data.style, data.staticStyle]"
    v-bind="data.attrs"
    v-on="listeners">
    <div class="label">
      <template v-if="props.label">{{ props.label }}</template>
      <slot name="label" />
    </div>
    <div class="value">
      <template v-if="props.value">{{ props.value }}</template>
      <slot />
    </div>
  </section>
</template>

<script>
/* eslint-disable vue/no-unused-properties -- https://github.com/vuejs/eslint-plugin-vue/issues/1312 */

export default {
  props: {
    name: {
      type: String,
      required: false,
      default: null,
    },
    label: {
      type: String,
      required: false,
      default: null,
    },
    value: {
      type: String,
      required: false,
      default: null,
    },
  },
};
</script>

```

**What did you expect to happen?**
No error, since `data`, `props` and `listeners` are part of the functional component context: https://vuejs.org/v2/guide/render-function.html#Functional-Components

**What actually happened?**
```
error: 'data' is not defined (vue/no-undef-properties) at ui/components/LabeledValue.vue:3:11:
  1 | <template functional>
  2 |   <section
> 3 |     :ref="data.ref"
    |           ^
  4 |     :class="[data.class, data.staticClass, props.name]"
  5 |     :style="[data.style, data.staticStyle]"
  6 |     v-bind="data.attrs"


error: 'props' is not defined (vue/no-undef-properties) at ui/components/LabeledValue.vue:4:44:
  2 |   <section
  3 |     :ref="data.ref"
> 4 |     :class="[data.class, data.staticClass, props.name]"
    |                                            ^
  5 |     :style="[data.style, data.staticStyle]"
  6 |     v-bind="data.attrs"
  7 |     v-on="listeners">


error: 'listeners' is not defined (vue/no-undef-properties) at ui/components/LabeledValue.vue:7:11:
   5 |     :style="[data.style, data.staticStyle]"
   6 |     v-bind="data.attrs"
>  7 |     v-on="listeners">
     |           ^
   8 |     <div class="label">
   9 |       <template v-if="props.label">{{ props.label }}</template>
  10 |       <slot name="label" />
```

**Repository to reproduce this issue**

[OpenLightingProject/open-fixture-library@`debug-eslint-vue-7.20.0`/ui/components/LabeledValue.vue](https://github.com/OpenLightingProject/open-fixture-library/blob/debug-eslint-vue-7.20.0/ui/components/LabeledValue.vue)

Note that the ESLint config file there is more elaborate than the minimal example provided in this issue.

**Related issue:** #1312

### PR DESCRIPTION
**Pull request by ota-meshi**
close #1312
close #1669

### CODE REVIEW

File: lib/utils/property-references.js [hunk 1]
```diff
@@ -96,6 +96,12 @@ function definePropertyReferenceExtractor(context) {
   /** @type {{ toRefNodes: Set<ESNode>, toRefsNodes: Set<ESNode>} | null} */
   let toRefSet = null
 
+  let isFunctionalTemplate = false
+  const templateBody = context.getSourceCode().ast.templateBody
+  if (templateBody) {
+    isFunctionalTemplate = utils.hasAttribute(templateBody, 'functional')
+  }
```
Comment (FloEdelmann) [inline lib/utils/property-references.js:103]: This doesn't catch this case:

```vue
<template>…</template>

<script>
export default {
  functional: true,
  …
}
</script>
```

Please also add test cases for this.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Very cool! I have one little remark though.

File: lib/utils/property-references.js (hunk 1 above)
Comment (ota-meshi) [reply lib/utils/property-references.js:103]: Hmm... I can't get it to work well using `functional: true` in SFC. Should it work?
(I've never used functional component before so I could be wrong.)

Comment (FloEdelmann) [reply lib/utils/property-references.js:103]: Functional components without a template, but with a render function are initialized that way. See https://vuejs.org/v2/guide/render-function.html#Functional-Components

Comment (ota-meshi) [reply lib/utils/property-references.js:103]: Thank you for the explanation. I understand. I think the render function is already supported.

https://github.com/vuejs/eslint-plugin-vue/blob/ae34c6582981ffc31ec03176f4fc417ba2f2412d/tests/lib/rules/no-undef-properties.js#L198-L204

Comment (FloEdelmann) [reply lib/utils/property-references.js:103]: Alright, then this fine for me.

**Review by FloEdelmann — APPROVED** [review]

**Comment by adamchipperfield** [comment]
I'm running 8.7.1 and am still having this issue, but it does look like it was released. Any ideas?

**Comment by adamchipperfield** [comment]
> I'm running 8.7.1 and am still having this issue, but it does look like it was released. Any ideas?

On deeper investigation it doesn't detect the props used in the **template** of a SFC. For example, the prop `handle` exists and is used in `<template>` via `props.handle`. But the linter still calls out `handle` as not being used.

**Pull request merged**